File: WASMiOS/WASMiOS/InterfaceTestController.m
//
//  InterfaceTestController.m
//  WASMiOS
//
//  Created by Pawel Grzmil on 17/01/16.
//  Copyright © 2016 Pawel Grzmil. All rights reserved.
//

#import "InterfaceTestController.h"
#import "InterfaceTestCell.h"

@interface InterfaceTestController()<UITableViewDataSource, UITableViewDelegate>
@property (weak, nonatomic) IBOutlet UILabel *fpsLabel;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic, strong) CADisplayLink *displayLink;
@property (nonatomic) CFTimeInterval startTime;
@property (nonatomic) NSInteger frameCount;
@property (nonatomic) NSArray* items;

@end;
@implementation InterfaceTestController
static NSString *simpleTableIdentifier = @"TableCell";

-(void)viewDidLoad{
    [super viewDidLoad];
    self.title = @"Test interfejsu";
    self.items = [NSArray arrayWithObjects:@"Vegetables", @"Fruits", @"Flower Buds", @"Legumes", @"Bulbs", @"Tubers",@"Vegetables", @"Fruits", @"Flower Buds", @"Legumes", @"Bulbs", @"Tubers",@"Vegetables", @"Fruits", @"Flower Buds", @"Legumes", @"Bulbs", @"Tubers",@"Vegetables", @"Fruits", @"Flower Buds", @"Legumes", @"Bulbs", @"Tubers",@"Vegetables", @"Fruits", @"Flower Buds", @"Legumes", @"Bulbs", @"Tubers",@"Vegetables", @"Fruits", @"Flower Buds", @"Legumes", @"Bulbs", @"Tubers", nil];
    [self startDisplayLink];
    
    }

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.items.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TableCell"];
    
    if (cell == nil) {
        cell = [[InterfaceTestCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"TableCell"];
    }
    
    [((InterfaceTestCell*)cell) updateCell:self.items[indexPath.row]];
    return cell;
}

- (void)startDisplayLink
{
    self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(handleDisplayLink:)];
    self.startTime = CACurrentMediaTime();
    [self.displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
}

- (void)stopDisplayLink
{
    [self.displayLink invalidate];
    self.displayLink = nil;
}

- (void)handleDisplayLink:(CADisplayLink *)displayLink
{
    self.frameCount++;
    
    CFTimeInterval now = CACurrentMediaTime();
    CFTimeInterval elapsed = now - self.startTime;
    
    if (elapsed >= 1.0) {
        CGFloat frameRate = self.frameCount / elapsed;
        self.fpsLabel.text = [NSString stringWithFormat:@"%.0f fps", frameRate];
        
        self.frameCount = 0;
        self.startTime = now;
    }
}
@end

File: Xamarin/XamarinResearch/Xamarin.iOS/ViewControllers/InterfaceTestController.cs
using CoreAnimation;
using CoreGraphics;
using Foundation;
using System;

using UIKit;

namespace Xamarin.iOS.ViewControllers
{
    public partial class InterfaceTestController : UIViewController
    {
        public int frameCount { get; set; }
        public CADisplayLink displayLink { get; set; }
        public double startTime { get; set; }

        public InterfaceTestController(IntPtr handle) : base(handle)
        {
        }

        public override void DidReceiveMemoryWarning()
        {
            base.DidReceiveMemoryWarning();
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            Title = "Test interfejsu";
            string[] tableItems = new string[] { "Vegetables", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Tubers", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Vegetables", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Tubers", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs", "Fruits", "Flower Buds", "Legumes", "Bulbs" };
            tableView.Source = new TableSource(tableItems);
            startFpsCounter();
        }

        void startFpsCounter()
        {
            displayLink = CADisplayLink.Create(updateFpsCounter);
            startTime = CAAnimation.CurrentMediaTime();
            displayLink.AddToRunLoop(NSRunLoop.Current, NSRunLoopMode.Common);
        }

        void stopFpsCounter()
        {
            displayLink.Invalidate();
            displayLink = null;
        }

        void updateFpsCounter()
        {
            frameCount++;
            double now = CAAnimation.CurrentMediaTime();
            double elapsed = now - startTime;

            if (elapsed >= 1.0)
            {
                double frameRate = this.frameCount / elapsed;
                fpsLabel.Text = string.Format("{0} fps", Math.Round(frameRate));
                frameCount = 0;
                startTime = now;
            }
        }

    }
}